filebeat/input/file/states.go

```diff
@@ -0,0 +1,155 @@
+package file
+
+import (
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// States handles list of FileState. One must use NewStates to instantiate a
+// file states regisry. Using the zero-value is not safe.
+type States struct {
+	sync.RWMutex
+
+	// states store
+	states []State
+
+	// idx maps state IDs to state indexes for fast lookup and modifications.
+	idx map[string]int
+}
+
+// NewStates generates a new states registry.
+func NewStates() *States {
+	return &States{
+		states: nil,
+		idx:    map[string]int{},
+	}
+}
+
+// Update updates a state. If previous state didn't exist, new one is created
+func (s *States) Update(newState State) {
+	s.UpdateWithTs(newState, time.Now())
+}
+
+// UpdateWithTs updates a state, assigning the given timestamp.
+// If previous state didn't exist, new one is created
+func (s *States) UpdateWithTs(newState State, ts time.Time) {
+	s.Lock()
+	defer s.Unlock()
+
+	id := newState.ID()
+	index := s.findPrevious(id)
+	newState.Timestamp = ts
+
+	if index >= 0 {
+		s.states[index] = newState
+	} else {
+		// No existing state found, add new one
+		s.idx[id] = len(s.states)
+		s.states = append(s.states, newState)
+		logp.Debug("input", "New state added for %s", newState.Source)
+	}
+}
+
+// FindPrevious lookups a registered state, that matching the new state.
+// Returns a zero-state if no match is found.
+func (s *States) FindPrevious(newState State) State {
+	s.RLock()
+	defer s.RUnlock()
+	i := s.findPrevious(newState.ID())
+	if i < 0 {
+		return State{}
+	}
+	return s.states[i]
+}
+
+// findPrevious returns the previous state fo the file
+// In case no previous state exists, index -1 is returned
+func (s *States) findPrevious(id string) int {
+	if i, exists := s.idx[id]; exists {
+		return i
+	}
+	return -1
+}
+
+// Cleanup cleans up the state array. All states which are older then `older` are removed
+// The number of states that were cleaned up and number of states that can be
+// cleaned up in the future is returned.
+
+func (s *States) Cleanup() (int, int) {
+	s.Lock()
+	defer s.Unlock()
+
+	currentTime := time.Now()
+	statesBefore := len(s.states)
+	numCanExpire := 0
+
+	for i := 0; i < len(s.states); i++ {
+		state := &s.states[i]
+		expired := (state.TTL > 0 && currentTime.Sub(state.Timestamp) > state.TTL)
+
+		if state.TTL == 0 || expired {
+			if !state.Finished {
+				logp.Err("State for %s should have been dropped, but couldn't as state is not finished.", state.Source)
+				continue
+			}
+
+			// remove state from index
+			delete(s.idx, state.ID())
+
+			// remove entry by relocating last entry + shrink states array
+			last := len(s.states) - 1
+			s.states[i] = s.states[last]
+			s.states = s.states[:last]
```

Comment:
thinking out loud here, related to the case of a lot of files.
Could this generate a lot of eviction/resizing in the real world? I guess we are just moving pointers.

Reply:
You mean CPU cache? No idea without testing/profiling. But we basically narrow down the array from beginning and end at the same time.

Reply:
Also depends on how many entries need to be removed.